### PR TITLE
PoC: Simplifying DTO Validation with Zod

### PR DESCRIPTION
**Pull request by timonrieger**
## Overview

This PR is the first iteration of a proposal to migrate from decorator-based DTO validation to a dedicated validation library: **Zod**.  
The idea has previously been discussed in Discord.

This is **not** a full migration of the server codebase. It focuses only on a subset of endpoints - specifically, the activity endpoints.

Because of this partial scope, both the current and proposed validation approaches must coexist in some areas. For example:

- `UserResponseSchema` exists alongside `UserResponseDto`.

After a full rollout of this PoC, the decorator-based approach would be removed entirely.

Additional annotations are included in the PR to explain key changes and subtle implementation details.

If this direction is accepted, the plan is to continue the migration in multiple follow-up PRs.

## Goals

- No breaking changes to the OpenAPI spec
- Preserve existing validation behavior
- Improve developer experience (DX) for data validation
- Improve type safety and IDE autocomplete
- Reduce code duplication
- Remove custom boilerplate (e.g., `validation.ts`)

### 1. Single Source of Truth  
The schema defines:
- Types  
- Validation rules  
- Transformations  
- OpenAPI metadata  

All in one place, instead of splitting concerns across decorators and helpers.

### 2. Less Framework Glue  
- No reflection metadata  
- Fewer custom decorators  
- No `class-transformer` quirks  
- Fewer moving parts → fewer edge cases  

### 3. Better Reuse & Testability  
Schemas are plain objects:
- Easy to compose
- Easy to share across layers
- Easy to test without spinning up Nest

## References
- [Zod Docs](https://zod.dev/)
- [NestJS-Zod Docs](https://github.com/BenLorantfy/nestjs-zod)

### CODE REVIEW

File: open-api/bin/generate-open-api.sh [hunk 1]
```diff
@@ -15,7 +15,7 @@ function dart {
   patch --no-backup-if-mismatch -u api.mustache <api.mustache.patch
 
   cd ../../
-  pnpm dlx @openapitools/openapi-generator-cli generate -g dart -i ./immich-openapi-specs.json -o ../mobile/openapi -t ./templates/mobile
+  pnpm dlx @openapitools/openapi-generator-cli generate -g dart -i ./immich-openapi-specs.json -o ../mobile/openapi -t ./templates/mobile --skip-validate-spec
```
Comment (timonrieger) [inline open-api/bin/generate-open-api.sh]: this flag is needed as the OpenApi spec generated via the nestjs-zod library does not reuse schemas by default (but inlines them in each reference), so `meta: { id: ... }` has to be used to achieve the existing behavior. This causes the schema in the spec to include the `id` field (the schema identifier) which causes validation errors when running the openapi-generator-cli. This flag does not cause any code generation changes and is only behavior related.

see
https://github.com/immich-app/immich/blob/d5a43e07bd30198fbd329a281ac51d5feb873c48/server/src/dtos/activity.dto.ts#L15
and
https://github.com/immich-app/immich/blob/d5a43e07bd30198fbd329a281ac51d5feb873c48/open-api/immich-openapi-specs.json#L20791

File: open-api/typescript-sdk/src/fetch-client.ts [hunk 2]
```diff
@@ -14,9 +14,9 @@ const oazapfts = Oazapfts.runtime(defaults);
 export const servers = {
     server1: "/api"
 };
-export type UserResponseDto = {
+export type UserResponseDtoOutput = {
```
Comment (timonrieger) [inline open-api/typescript-sdk/src/fetch-client.ts:17]: This DTO has not been renamed, it is still defined a few lines below. 

As noted in the PR description the duplication of this schema (note the `Output` suffix) is a side effect of having to maintain both approaches in this PR, causing the generation of two models for the user response model. As mentioned this behavior is only temporary during the gradual migration and will revert to status quo after completion.

File: open-api/typescript-sdk/src/fetch-client.ts [hunk 3]
```diff
@@ -6822,6 +6810,18 @@ export enum NotificationType {
...
+    Pink = "pink",
+    Red = "red",
+    Yellow = "yellow",
+    Blue = "blue",
+    Green = "green",
+    Purple = "purple",
+    Orange = "orange",
+    Gray = "gray",
+    Amber = "amber"
+}
```
Comment (timonrieger) [inline open-api/typescript-sdk/src/fetch-client.ts:6848]: same applies here

File: open-api/immich-openapi-specs.json [hunk 4]
```diff
@@ -25576,6 +25485,61 @@
         ],
         "type": "object"
       },
+      "UserResponseDto_Output": {
```
Comment (timonrieger) [inline open-api/immich-openapi-specs.json:25587]: see comments regarding the TS client above

File: server/src/controllers/activity.controller.ts [hunk 5]
```diff
@@ -28,7 +29,8 @@ export class ActivityController {
       'Returns a list of activities for the selected asset or album. The activities are returned in sorted order, with the oldest activities appearing first.',
     history: new HistoryBuilder().added('v1').beta('v1').stable('v2'),
   })
-  getActivities(@Auth() auth: AuthDto, @Query() dto: ActivitySearchDto): Promise<ActivityResponseDto[]> {
+  @ZodResponse({ type: [ActivityResponseDto], status: HttpStatus.OK })
+  getActivities(@Auth() auth: AuthDto, @Query() dto: ActivitySearchDto) {
```
Comment (timonrieger) [inline server/src/controllers/activity.controller.ts]: this is the approach for annotating responses with their respective models.

File: server/src/dtos/activity.dto.ts [hunk 6]
```diff
@@ -1,84 +1,107 @@
...
-
-export class ActivityStatisticsResponseDto {
-  @ApiProperty({ type: 'integer', description: 'Number of comments' })
-  comments!: number;
-
-  @ApiProperty({ type: 'integer', description: 'Number of likes' })
-  likes!: number;
-}
+// Option 2 - define inline and access via .enum.XXX:
+export const ReactionType = z.enum(['comment', 'like']).describe('Reaction type').meta({ id: 'ReactionType' });
```
Comment (timonrieger) [inline server/src/dtos/activity.dto.ts]: two different approaches for handling enums. This is a server-only change, non-breaking for clients.

The first would integrate quickly with the current code as we can keep the existing enums in `enum.ts` as is. However it is more verbose since we need to create a schema for each enum.

The second approach needs some refactoring in existing enum references, but is less verbose. The refactor looks like this:

```diff
-     if (dto.type === ReactionType.LIKE) {
+     if (dto.type === ReactionType.enum.like) {
```
https://github.com/immich-app/immich/blob/042bbcae8bfac1fde48c28bd629f92017c891e41/server/src/services/activity.service.ts#L49

**I tend towards the second approach.**

[Read more about zod enums](https://zod.dev/api?id=enums)

File: server/src/dtos/activity.dto.ts [hunk 7]
```diff
@@ -1,84 +1,107 @@
...
+const ActivityResponseSchema = z
+  .object({
+    id: z.uuidv4().describe('Activity ID'),
+    createdAt: z.iso.datetime().describe('Creation date'),
+    user: UserResponseSchema,
+    assetId: z.uuidv4().describe('Asset ID (if activity is for an asset)').nullable(),
+    type: ReactionType,
+    comment: z.string().describe('Comment text (for comment activities)').nullish(),
+  })
+  .describe('Activity response');
```
Comment (timonrieger) [inline server/src/dtos/activity.dto.ts:33]: this is the core change proposed in this PR. Moving away from decorator to a schema approach

File: server/src/dtos/activity.dto.ts [hunk 8]
```diff
@@ -1,84 +1,107 @@
...
+  comment: z.string().describe('Comment text (required if type is comment)').optional(),
+})
+  .refine((data) => data.type !== ReactionType.enum.comment || (data.comment && data.comment.trim() !== ''), {
+    message: 'Comment is required when type is COMMENT',
+    path: ['comment'],
+  })
+  .refine((data) => data.type === ReactionType.enum.comment || !data.comment, {
+    message: 'Comment must not be provided when type is not COMMENT',
+    path: ['comment'],
+  })
```
Comment (timonrieger) [inline server/src/dtos/activity.dto.ts]: this is value the current approach cannot provide. Validating the data input depending on conditionals. 

This could look like this on a request:

Image
Image

File: server/src/dtos/activity.dto.ts [hunk 9]
```diff
@@ -1,84 +1,107 @@
...
+    },
+    comment: null,
   };
 };
+
+export class ActivityResponseDto extends createZodDto(ActivityResponseSchema) {}
+export class ActivityCreateDto extends createZodDto(ActivityCreateSchema) {}
+export class ActivityDto extends createZodDto(ActivitySchema) {}
+export class ActivitySearchDto extends createZodDto(ActivitySearchSchema) {}
+export class ActivityStatisticsResponseDto extends createZodDto(ActivityStatisticsResponseSchema) {}
```
Comment (timonrieger) [inline server/src/dtos/activity.dto.ts:111]: to reduce refactoring hell and ensuring framework compatibility, `nestjs-zod` has a util to create NestJS DTOs from zod schemas. Since classes are limited in expressing complex union types (implemented with e.g. `zod.coerce`), they don't work on all schemas. Thus, above the `zod.refine` util has the same validation effect by avoiding the type complexity for the class conversion.

Using this approach existing references to the Dto remain intact as is.

File: server/src/dtos/user.dto.ts [hunk 10]
```diff
@@ -51,6 +52,15 @@ export class UserResponseDto {
...
 }
 
+export const UserResponseSchema = z.object({
+  id: z.uuidv4().describe('User ID'),
+  name: z.string().describe('User name'),
+  email: z.string().describe('User email'),
+  profileImagePath: z.string().describe('Profile image path'),
+  avatarColor: z.enum(UserAvatarColor).describe('Avatar color'),
+  profileChangedAt: z.iso.datetime().describe('Profile change date'),
+}).meta({ id: 'UserResponseDto' });
```
Comment (timonrieger) [inline server/src/dtos/user.dto.ts:62]: the meta object ensures the schema is being reused and does not conflict with others when generating the OpenApi spec.

File: server/src/middleware/global-exception.filter.ts [hunk 11]
```diff
@@ -33,6 +35,13 @@ export class GlobalExceptionFilter implements ExceptionFilter<Error> {
     logGlobalError(this.logger, error);
 
     if (error instanceof HttpException) {
+      // Handle ZodSerializationException
+      if (error instanceof ZodSerializationException) {
+        const zodError = error.getZodError();
+        if (zodError instanceof ZodError) {
+          // TODO: handle zod error
```
Comment (timonrieger) [inline server/src/middleware/global-exception.filter.ts]: tbd, the current validation error response without custom processing looks like this

```json
{
  "statusCode": 400,
  "message": "Validation failed",
  "errors": [
     {
        "code": "custom",
        "path": ["comment"],
        "message": "Comment is required when type is COMMENT",
    },
  ],
  "correlationId": "36rcgn1a"
}
```

File: server/src/services/activity.service.spec.ts [hunk 12]
```diff
@@ -35,7 +35,7 @@ describe(ActivityService.name, () => {
       mocks.activity.search.mockResolvedValue([]);
 
       await expect(
-        sut.getAll(factory.auth({ user: { id: userId } }), { assetId, albumId, type: ReactionType.LIKE }),
+        sut.getAll(factory.auth({ user: { id: userId } }), { assetId, albumId, type: ReactionType.enum.like }),
```
Comment (timonrieger) [inline server/src/services/activity.service.spec.ts]: as mentioned above, the refactor when going for the second approach handling enums.

File: server/src/utils/misc.ts [hunk 13]
```diff
@@ -162,10 +200,42 @@ const patchOpenAPI = (document: OpenAPIObject) => {
...
+        continue;
+      }
+
+      if (_.isEqual(stripSchemaMetadata(baseSchema), stripSchemaMetadata(outputSchema))) {
+        delete schemas[outputName];
+        schemaNameMap[outputName] = baseName;
+      }
+    }
+
+    replaceSchemaRefs(document, schemaNameMap);
```
Comment (timonrieger) [inline server/src/utils/misc.ts:234]: nestjs-zod emits separate "_Output" component schemas for response types
(via z.toJSONSchema's "output" IO mode). This is correct when input/output
shapes genuinely differ (e.g. coercion, transforms, codecs), but for most
schemas (especially enums) input and output are identical, producing
redundant schemas that cause client generators to emit duplicate models
(e.g. ReactionType + ReactionType_Output, ActivityResponseDto + ActivityResponseDto_Output).

Collapse "_Output" variants that are identical to their base schema, and
rename orphaned "_Output" schemas (where no base exists) to keep generated
client model names stable and free of unnecessary duplication.

**This openapi patch is necessary, as it prevents model duplication, response type unions, and structural divergences that could break existing client code across clients (TS, Python, Dart)**

**Comment by timonrieger** [comment]
as a side note, [orval](https://orval.dev/) allows reversing the openapi spec to zod schemas. This makes this implementations way easier and reliable

save this as `orval.config.ts`, download the current immich OpenApi spec to `spec.json`, then run `pnpm dlx --package zod --package orval orval`.
```ts
import { defineConfig } from 'orval';

export default defineConfig({
  immich: {
    output: {
      client: 'zod',
      mode: 'single',
      target: './src/api/immich.schemas.ts',
    },
    input: {
      target: './spec.json',
    },
  },
});
```

File: open-api/bin/generate-open-api.sh (hunk 1 above)
Comment (jrasm91) [reply open-api/bin/generate-open-api.sh]: I'm not a huge fan of this as a solution, since it can catch meaningful issues in the specification file.

File: server/src/controllers/activity.controller.ts (hunk 5 above)
Comment (jrasm91) [reply server/src/controllers/activity.controller.ts]: So it is completely disconnected from the response type?

File: server/src/dtos/activity.dto.ts (hunk 6 above)
Comment (jrasm91) [reply server/src/dtos/activity.dto.ts]: I'd probably vote for option one.

File: server/src/dtos/activity.dto.ts (hunk 8 above)
Comment (jrasm91) [reply server/src/dtos/activity.dto.ts]: We do already have and do this, so I'm not sure why you are saying we don't or can't.

File: server/src/services/activity.service.spec.ts (hunk 12 above)
Comment (jrasm91) [reply server/src/services/activity.service.spec.ts]: Probably worth keeping the first one, so we don't have to rewrite all usages of enums through out the codebase.

File: server/src/utils/misc.ts (hunk 13 above)
Comment (jrasm91) [reply server/src/utils/misc.ts:234]: Why does it even generate these _Output types in the first place? All of our schemas are already explicitly for requests or responses, not both at the same time.

Comment (timonrieger) [reply server/src/utils/misc.ts:234]: `nestjs-zod` generates `_Output` types because it treats Zod schemas as potentially having different input and output shapes, even when we already separate request and response DTOs.

It does this because Zod schemas can have different input vs output types due to:
- `.transform()` 
- `.preprocess()`
- `.coerce.*`
- `.default()` (affects required/optional)
- Codecs (encode vs decode)

So it generates both variants to be safe, even if the schema doesn't use any of these features.

File: server/src/controllers/activity.controller.ts (hunk 5 above)
Comment (timonrieger) [reply server/src/controllers/activity.controller.ts]: no, not at all actually. It's a single source of truth that keeps compile-time types, runtime serialization, and API docs all in sync. check the library docs: https://github.com/BenLorantfy/nestjs-zod/blob/main/packages/nestjs-zod/src/response.ts#L16:L16

File: open-api/bin/generate-open-api.sh (hunk 1 above)
Comment (timonrieger) [reply open-api/bin/generate-open-api.sh]: i just pushed a commit to remove the flag and rather strip the id field from the generated spec to preserve the validation capability 5621cf9fcd69179bcdbf74115c77e47942b3c47b

File: server/src/dtos/activity.dto.ts (hunk 8 above)
Comment (timonrieger) [reply server/src/dtos/activity.dto.ts]: partially, it does not raise when a comment is supplied on `type: like`, but silently strips it, when testing against demo

image


image

File: server/src/utils/misc.ts (hunk 13 above)
Comment (jrasm91) [reply server/src/utils/misc.ts:234]: That doesn't seem ideal, based on our usage. Can we turn that off or automatically always remove them?

Comment (timonrieger) [reply server/src/utils/misc.ts:234]: as far as i researched, we cannot turn this off as this is core behavior for zod. The current approach is to patch the spec to clean this up, so the naming remains as is

File: server/src/middleware/global-exception.filter.ts (hunk 11 above)
Comment (timonrieger) [reply server/src/middleware/global-exception.filter.ts]: decided to maintain existing error schema:
```json
{
  "message": [
    "Comment must not be provided when type is not COMMENT"
  ],
  "error": "Bad Request",
  "statusCode": 400,
  "correlationId": "dtw6imvq"
}
```

**Comment by timonrieger** [comment]
proposal is discussed, see implementation in #26597
